fix(run): Ensure 'run' subcommand works with host proxy settings.

melodyliu1986 opened the pull request:
When 'ramalama run' is used with '--network none', it was inheriting host proxy environment variables. This caused the interanl client to fail when connecting to the internal llama-server on 127.0.0.1, as it tried to route loopback traffic through the unreachable proxy.

This change modifies engine.py to:
- Correctly set NO_PROXY/no_proxy for localhost and 127.0.0.1.
- Explicitly unset http_proxy, https_proxy, HTTP_PROXY, and HTTPS_PROXY variables for the container when the 'run' subcommand is invoked.

This allows the internal client to connect directly to the internal server, resolving the connection error.

Fixes: #1414

**Testings before the change:**
```
$ env|grep "proxy"
no_proxy=localhost,127.0.0.0/8,::1
https_proxy=http://proxy4.squi-001.prod.rdu2.dc.redhat.com:3128/
HTTPS_PROXY=http://proxy4.squi-001.prod.rdu2.dc.redhat.com:3128/
HTTP_PROXY=http://proxy4.squi-001.prod.rdu2.dc.redhat.com:3128/
http_proxy=http://proxy4.squi-001.prod.rdu2.dc.redhat.com:3128/

$ ramalama run granite3-moe
🦭 > how are you?
Error: could not connect to: http://127.0.0.1:8080/v1/chat/completions

```

**Testings after the change:**
```
$ env|grep "proxy"
no_proxy=localhost,127.0.0.0/8,::1
https_proxy=http://proxy4.squi-001.prod.rdu2.dc.redhat.com:3128/
HTTPS_PROXY=http://proxy4.squi-001.prod.rdu2.dc.redhat.com:3128/
HTTP_PROXY=http://proxy4.squi-001.prod.rdu2.dc.redhat.com:3128/
http_proxy=http://proxy4.squi-001.prod.rdu2.dc.redhat.com:3128/

$ ramalama list
NAME                                      MODIFIED   SIZE     
ollama://granite3-moe/granite3-moe:latest 5 days ago 783.77 MB
ollama://tinyllama/tinyllama:latest       5 days ago 608.16 MB

$ ramalama run granite3-moe
🦭 > how are you?
I'm an artificial intelligence, so I don't have feelings, but I'm here and ready to assist you with any questions you have. How can I help you today?
🦭 > what is the longest river and highest montain in the world?
1. The longest river in the world is the Nile River, which runs through Africa. It is approximately 4,135 miles (6,650 kilometers) long.

2. The highest mountain in the world is Mount Everest. It is an extinct volcano located in the Himalayas on the border of Nepal and Tibet. It stands at an elevation of approximately 29,029 feet (8,848 meters) above sea level.
🦭 > 

$ ramalama serve granite3-moe
serving on port 8080
build: 5335 (d8919424) with cc (GCC) 11.5.0 20240719 (Red Hat 11.5.0-5) for x86_64-redhat-linux
system info: n_threads = 4, n_threads_batch = 4, total_threads = 8

system_info: n_threads = 4 (n_threads_batch = 4) / 8 | CPU : SSE3 = 1 | SSSE3 = 1 | AVX = 1 | AVX2 = 1 | F16C = 1 | FMA = 1 | BMI2 = 1 | LLAMAFILE = 1 | OPENMP = 1 | AARCH64_REPACK = 1 | 

main: binding port with default address family
main: HTTP server is listening, hostname: 0.0.0.0, port: 8080, http threads: 7

```

## Summary by Sourcery

Ensure 'run' subcommand bypasses host proxy settings to allow internal client to connect to the local llama server

Bug Fixes:
- Prevent host proxy environment variables from interfering with internal loopback connections by unsetting http_proxy, https_proxy, HTTP_PROXY, and HTTPS_PROXY in the 'run' subcommand
- Extend NO_PROXY/no_proxy to include localhost and 127.0.0.1 for containers when 'run' is invoked

Enhancements:
- Include debug flag in the image pull step and emit a debug-only note if the pull command fails

ericcurtin commented:
Could we put all this in libexec/ramalama/ramalama-run-core and instead? It's cleaner no if statement required since at this point we know we are in the run code for sure

rhatdan commented:
Would it be better/easier to just set podman run --network=none --http_proxy=false?

Although this command does not exist with Docker.

melodyliu1986 commented:
> The thing is if we put it in libexec/ramalama/ramalama-run-core and solve it in python it should work with docker/podman/nocontainers

I reverted the engine.py and switched to libexec/ramalama/ramalama-run-core, can you please review it? TY

rhatdan commented:
Since the ramalama-run-core is happening in a container, then are we sure the environment variables from the host are getting passed into the command?

rhatdan commented:
Yes we rebuild on Mondays usually, so once you remove those two lines I pointed out, I will merge and this fix should show up in next release.

rhatdan commented:
Thanks @melodyliu1986

ericcurtin commented:
> Since the ramalama-run-core is happening in a container, then are we sure the environment variables from the host are getting passed into the command?

Thing is, ramalama-run-core should never use a proxy, it should always be localhost. ramalama-client-core called without ramalama-run-core could potentially use a proxy for a remote server.
